proto/pulumi/provider.proto

```diff
@@ -363,81 +373,135 @@ message CheckFailure {
     string reason = 2;
 }
 
+// `DiffRequest` is the type of requests sent as part of [](pulumirpc.ResourceProvider.DiffConfig) and
+// [](pulumirpc.ResourceProvider.Diff) calls. A `DiffRequest` primarily captures:
+//
+// * the URN of the resource whose properties are being compared;
+// * the old and new input properties of the resource; and
+// * the old output properties of the resource.
+//
+// In the case of [](pulumirpc.ResourceProvider.DiffConfig), the URN will be the URN of the provider resource being
+// examined, which may or may not be a [default provider](default-providers), and the inputs and outputs will be the
+// provider configuration and state. Inputs supplied to a [](pulumirpc.ResourceProvider.DiffConfig) call should have
+// been previously checked by a call to [](pulumirpc.ResourceProvider.CheckConfig); inputs supplied to a
+// [](pulumirpc.ResourceProvider.Diff) call should have been previously checked by a call to
+// [](pulumirpc.ResourceProvider.Check).
 message DiffRequest {
-    string id = 1;                     // the ID of the resource to diff.
-    string urn = 2;                    // the Pulumi URN for this resource.
-    google.protobuf.Struct olds = 3;   // the old output values of resource to diff.
+    // The ID of the resource being diffed.
+    string id = 1;
 
-    // the new input values of resource to diff, copied from CheckResponse.inputs.
+    // The URN of the resource being diffed.
+    string urn = 2;
+
+    // The old *output* properties of the resource being diffed.
+    google.protobuf.Struct olds = 3;
+
+    // The new *input* properties of the resource being diffed. These should have been validated by an appropriate call
+    // to [](pulumirpc.ResourceProvider.CheckConfig) or [](pulumirpc.ResourceProvider.Check).
     google.protobuf.Struct news = 4;
 
-    repeated string ignoreChanges = 5; // a set of property paths that should be treated as unchanged.
-    google.protobuf.Struct old_inputs = 6;   // the old input values of the resource to diff.
-    string name = 7; // the Pulumi name for this resource.
-    string type = 8; // the Pulumi type for this resource.
+    // A set of [property paths](property-paths) that should be treated as unchanged.
```

Comment:
This is very confusing to have it in here. It is difficult to explain in-line but we could probably have an ingoreChanges page dedicated to the concept with the nitty gritty details. 

I tend to think of this in this way: the engine handles ignore changes for simple providers, but providers customizing detailed Diff might need to be aware of it to handle it even better.

Reply:
Tracking in #17691